qiskit/circuit/qpy_serialization.py

```diff
@@ -936,8 +1024,96 @@ def _write_instruction(file_obj, instruction_tuple, custom_instructions, index_m
         container.close()
 
 
+def _write_pauli_evolution_gate(file_obj, evolution_gate):
+    operator_list = evolution_gate.operator
+    standalone = False
+    if not isinstance(operator_list, list):
+        operator_list = [operator_list]
+        standalone = True
+    num_operators = len(operator_list)
+    pauli_data_buf = io.BytesIO()
+    for operator in operator_list:
+        element_buf = io.BytesIO()
+        buf = io.BytesIO()
+        pauli_list = operator.to_list(array=True)
+        np.save(buf, pauli_list)
+        data = buf.getvalue()
+        element_metadata = struct.pack(SPARSE_PAULI_OP_LIST_ELEM_PACK, len(data))
+        element_buf.write(element_metadata)
+        element_buf.write(data)
+        pauli_data_buf.write(element_buf.getvalue())
+    time = evolution_gate.time
+    if isinstance(time, float):
+        time_type = b"f"
+        time_data = struct.pack("!d", time)
+        time_size = struct.calcsize("!d")
+    elif isinstance(time, Parameter):
+        time_type = b"p"
+        buf = io.BytesIO()
+        _write_parameter(buf, time)
+        time_data = buf.getvalue()
+        time_size = len(time_data)
+    elif isinstance(time, ParameterExpression):
+        time_type = b"e"
+        buf = io.BytesIO()
+        _write_parameter_expression(buf, time)
+        time_data = buf.getvalue()
+        time_size = len(time_data)
+    else:
+        raise TypeError(f"Invalid time type {time} for PauliEvolutionGate")
+
+    synth_class = str(type(evolution_gate.synthesis).__name__)
```

Comment:
The `str` seems unnecessary.  I'm not 100% sure what this is being used for, but should it use the fully qualified name instead of just the class name in case of clashes?  (Though that will limit it to classes that posses a fully qualified name.)

Reply:
We need the class name here to build an equivalent object of the same class with the same parameters. Basically anything in https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/synthesis/__init__.py is a valid type here (although `EvolutionSynthesis` or `ProductFormula` are abstract there). While there could be custom subclasses too we're not supporting those in qpy. So I just need to be able to do `getattr(qiskit.synthesis, synth_class)` or equivalent. The str cast is probably unnecessary though I agree

Reply:
Sounds reasonable within the limitations to me.  If we extend it, we'll probably need the fully qualified name, but that's an issue for a different time.